IN-515  Replacing dummy text network endpoint with the actual implementation

adessy opened the pull request:
IN-515 (It must be reviewed after IN-474, bc it's based on it.)

Not entirely satisfied with this one. The bulk of the work is in:
- `Insights::FrontEndFormatTextNetwork` and 
- `NLP::TextNetwork`.

`NLP::TextNetwork` now has a method to merge multiple networks into a single network (bc we have one network for each locale on the BE side). To avoid collisions, between identifiers when merging networks, I introduced a distinction between the name and the id of the node and implemented a method to namespace the identifiers.

`Insights::FrontEndFormatTextNetwork` is computing the network representation needed by the FE. There are quite a few differences with the representation returned by the NLP service. For instance, in the FE representation:
- A node represents either a keyword (= nodes in the nlp representation) or a community.
- Links are not between keywords, but represent membership to a cluster or a link between clusters.

There are still a few unknowns left about how to compute (when merging networks) and normalise the values `val` of the FE representation.

adessy committed:
IN-515 Refactoring NLP::TextNetwork to allow merging networks

IN-515 #comment + check node ids when adding communities and links

IN-515 #time 0w 1d 1h 2m IN-515 Refactoring NLP::TextNetwork to allow merging networks

adessy committed:
IN-515 Add TextNetworkFrontEndFormatter service to help serializing networks in the format expected by the FE

IN-515 #time 0w 0d 5h 11m IN-515 Add TextNetworkFrontEndFormatter service to help serializing networks in the format expected by the FE

adessy committed:
IN-515 Add the network serializer

IN-515 #time 0w 0d 0h 1m Add the network serializer

adessy committed:
IN-515 Replace the dummy network endpoint

IN-515 #time 0w 0d 0h 1m Replace the dummy network endpoint

adessy committed:
IN-515 Rename TextNetworkFrontEndFormatter to FrontEndFormatTextNetwork

IN-515 #time 0w 0d 0h 14m Rename TextNetworkFrontEndFormatter to FrontEndFormatTextNetwork

adessy committed:
IN-515 Add comment

IN-515 #time 0w 0d 0h 2m Add comment

adessy committed:
IN-515 Fix (nlp) text network factory

IN-515 #time 0w 0d 0h 54m Fix (nlp) text network factory

adessy committed:
IN-515 Fix text-network equality

adessy committed:
IN-515 Use the pair of node ids to index links (instead of nested hashes)

IN-515 #time 0w 0d 0h 6m Use the pair of node ids to index links (instead of nested hashes)

adessy committed:
IN-515 Update documentation

IN-515 #time 0w 0d 0h 22m Update documentation

adessy committed:
IN-515 Add acceptances test for network endpoint

IN-515 #time 0w 0d 1h 9m Add acceptances test for network endpoint

adessy committed:
IN-515 Return 404 when there is no network yet

IN-515 #time 0w 0d 0h 1m Return 404 when there is no network yet

adessy committed:
IN-515 Aggregate failure in test

IN-515 #time 0w 0d 0h 2m Aggregate failure in test

adessy committed:
IN-515 Implement a shorter string representation for text networks

IN-515 #time 0w 0d 4h 33m Implement a shorter string representation for text networks

adessy committed:
IN-515 Add tests for the text-network FE adapter

IN-515 #comment + refactoring to make testing easier

IN-515 #time 0w 0d 0h 1m Add tests for the text-network FE adapter

adessy committed:
IN-515 Add documentation

IN-515 #comment IN-515 #time 0w 0d 1h 1m Add documentation

adessy committed:
IN-515 Use most important nodes to name cluster nodes

IN-515 #time 0w 0d 0h 14m Use most important nodes to name cluster nodes

SaraBCL commented on back/engines/commercial/insights/app/services/insights/front_end_format_text_network.rb:
```diff
@@ -0,0 +1,103 @@
+    def initialize(view)
+      @id = "network-#{view.id}"
+      @network = NLP::TextNetwork.merge(
+        # Namespacing networks wrt to the language to avoid id collisions.
```
Curious about this one, is it to avoid UUID collisions ? I thought they were very unlikely, plus the id is kept unique by the db constraint, maybe I'm misunderstanding something ?

adessy replied:
It's to avoid collisions between node identifiers in the FE representation of the network, not between record identifiers in the DB. The NLP service is identifying the node using the keyword. But in a multilingual context, the same keyword could appear in different languages and then they are no longer unique (a bit annoying for an identifier). 

Merging those nodes would have been another option, but:
- there is no guarantee that the same word in different languages refer to the same concept;
- it would make a lot of sense to combine ex post pagerank scores (importance score computed by the nlp service). 

So I opted to keep multiple nodes with the same keyword but different identifiers.

SaraBCL commented on back/engines/commercial/nlp/lib/nlp/text_network.rb:
```diff
@@ -1,22 +1,107 @@
+        links = networks.flat_map(&:links)
+        communities = networks.flat_map(&:communities)
+
+        raise 'Node identifier collision' if duplicates?(nodes.map(&:id))
```
Isn't the name-spacing avoiding this already ?

adessy replied:
Yep, but `merge` and `namespace` are somehow independent from each other. Here we are just making sure that the resulting network is valid and does not contain duplicates. So if u will, it checks if the networks were indeed appropriately namespaced before -- even if in practice, merge does not really care as long as there are no collisions.

SaraBCL reviewed (approved):
Neat. Once again nothing to nitpick, code-wise
My only concerns : 
- multilingual support, is the apposition meaningful ? Won't it simply be two networks, side by side, that we could have shown separately with a language switcher ?
- I don't know much about performance in rails, but I think merging the networks on the go each time we serve it, which is what I believe is happening here, could become problematic with big projects.

adessy commented:
Thanks for the review! Now, addressing your concerns 🙂 

> multilingual support, is the apposition meaningful ? Won't it simply be two networks, side by side, that we could have shown separately with a language switcher ?

Good point. Indeed, merged networks will be disconnected. Something to be discussed with product. For now, I just followed the current specification.

> I don't know much about performance in rails, but I think merging the networks on the go each time we serve it, which is what I believe is happening here, could become problematic with big projects.

To be seen 🙂 but personally not worried about that one, especially when we will limit the number of keywords (nodes). I think this will take (much?) less time than any request to the DB.

adessy committed:
Merge branch 'master' into IN-515/text-network-endpoint

IN-515 #time 0w 0d 2h 2m Merge branch 'master' into IN-515/text-network-endpoint
